filebeat/input/file/identifier.go

```diff
@@ -0,0 +1,219 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package file
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/mitchellh/hashstructure"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+const (
+	inodeDeviceIDName = "inode_deviceid"
+	pathName          = "path"
+	inodeMarkerName   = "inode_marker"
+
+	DefaultIdentifierName = inodeDeviceIDName
+	identitySep           = "::"
+)
+
+var (
+	identifierFactories = map[string]IdentifierFactory{
+		inodeDeviceIDName: newINodeDeviceIdentifier,
+		pathName:          newPathIdentifier,
+		inodeMarkerName:   newINodeMarkerIdentifier,
+	}
+)
+
+type IdentifierFactory func(*common.Config) (StateIdentifier, error)
+
+// StateIdentifier generates an ID for a State.
+type StateIdentifier interface {
+	// GenerateID generates and returns the ID of the state
+	GenerateID(*State)
```

Comment:
instead of modifying the state consider to turn the interface of the method into `GenerateID(State) (id, identifierType string)`.